packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,618 @@
+            _deposit(msg.sender, tokens[i].token, uint96(tokens[i].amount));
+        }
+
+        // Store the transfers so they can be processed later
+        _storeTransfers(transfers);
+    }
+
+    function beforeBlockSubmission(
+        bytes memory   txsData,
+        bytes calldata callbackData
+        )
+        external
+        onlyFromExchangeOwner
+    {
+        BridgeOperations memory operations = abi.decode(callbackData, (BridgeOperations));
+
+        Context memory ctx = Context({
+            domainSeparator: DOMAIN_SEPARATOR,
+            accountID: accountID,
+            txIdx: 0
+        });
+
+        _processTransfers(ctx, operations.transferBatches, txsData);
+        _processCalls(ctx, operations, txsData);
+
+        // Make sure we have consumed exactly the expected number of transactions
+        require(txsData.length == ctx.txIdx * ExchangeData.TX_DATA_AVAILABILITY_SIZE, "INVALID_NUM_TXS");
+    }
+
+    // Allows withdrawing from pending transfers that are at least MAX_AGE_PENDING_TRANSFERS old.
+    function withdrawFromPendingBatchDeposit(
+        uint                              batchID,
+        InternalBridgeTransfer[] calldata transfers,
+        uint[]                   calldata indices
+        )
+        external
+    {
+        // Check if withdrawing from these transfers is possible
+        bytes32 hash = _hashTransfers(batchID, transfers);
+        require(_arePendingTransfersTooOld(hash), "TRANSFERS_NOT_TOO_OLD");
+
+        for (uint i = 0; i < indices.length; i++) {
+            uint idx = indices[i];
+
+            require(!withdrawn[hash][idx], "ALREADY_WITHDRAWN");
+            withdrawn[hash][idx] = true;
+
+            address tokenAddress = exchange.getTokenAddress(transfers[idx].tokenID);
+
+            _transferOut(
+                tokenAddress,
+                transfers[idx].amount,
+                transfers[idx].owner
+            );
+        }
+    }
+
+    // Can be used to withdraw funds that were already deposited to the bridge,
+    // but need to be returned to be able to withdraw from old pending transfers.
+    function forceWithdraw(address[] calldata tokens)
+        external
+        payable
+    {
+        for (uint i = 0; i < tokens.length; i++) {
+            exchange.forceWithdraw{value: msg.value / tokens.length}(
+                address(this),
+                tokens[i],
+                accountID
+            );
+        }
+    }
+
+    function setConnectorTrusted(
+        address connector,
+        bool    trusted
+        )
+        external
+        onlyOwner
+    {
+        trustedConnectors[connector] = trusted;
+        emit ConnectorTrusted(connector, trusted);
+    }
+
+    receive()
+        external
+        payable
+    {}
+
+    // --- Internal functions ---
+
+    function _processTransfers(
+        Context         memory ctx,
+        TransferBatch[] memory operations,
+        bytes           memory txsData
+        )
+        internal
+    {
```

Comment:
Do we allow the corresponding deposit to be processed in a different block? If only the deposit is processed, but its L2 transfers are not, will users withdraw their tokens with `withdrawFromPendingBatchDeposit`? (in such case, the token is already transfered to DepositContract I guess).


If processing of deposit and transfers must be *atomic*, maybe we should also change `_hashTransfers` to hash the deposit data.

Reply:
Currently the deposit can be processed in a different block, completely separate to the L2 transfers.

In all cases funds are directly deposited to the deposit contract in `batchDeposit` so they have to be removed from there if the L2 transfers are not processed.
Two scenario's:
- Deposit itself not processed: funds can be withdrawn to the bridge by anyone with `Exchange.withdrawFromDepositRequest`.
- Deposit processed: funds can be withdrawn to the bridge by anyone with `Bridge.forceWithdraw`.

Once the funds are in the Bridge `withdrawFromPendingBatchDeposit` can be used to get the tokens back. So it's definitely not something we actually want normal users to have to go through (similar to how we always process all Exchange deposits), and so we should do all transfers in normal situations. But the biggest deposit cost is paid by the caller of `batchDeposit`, and we only have to do the cheaper L2 transfers, costing ~4,000 gas each. To help pay for that we can use up to `MAX_FEE_BIPS` of the deposited amount. But if people spam `batchDeposit` with very low value deposits we can choose not to process them.

Reply:
The above comment of yours answers some of my questions.
I wonder: if we process deposit and its L2 transfers in a single block **atomicly**, may we be able to completely remove `Bridge.forceWithdraw` to simplify the code? Or maybe from the DEX's perspecive, there is no way to guarantee the deposit is processed with the transfers all together as the deposit is a normal deposit and the DEX/operator can process it without even considering the rules in a Bridge. If the answer is NO, then we have to keep the current design and allow deposit/transfers to be processed in different blocks.

Reply:
You can do it atomicly, but that brings other problems:
- `batchDeposit` only does a transfer to the bridge. And for processing the deposits we also have to pay again to transfer funds to the deposit contract. So less efficient this way and more costs we have to pay (and recover fees for).
- Instead of moving tokens to the Deposit contract we could just keep the funds in the Bridge contract, and do "fake" deposit/withdrawals which just increase/decrease balances. Essentially this makes the Bridge contract a 2nd deposit contract where exchange funds could be stored. But that brings extra risk and I think overall more complexity.

Reply:
I don't like the *second deposit contract* idea, so lets keep the design as-is.